### PR TITLE
fix(encode): support arrays and label Option<T> as optional

### PR DESCRIPTION
**Pull request by ritvik-saronic**
### Changelog

- Added derive(Encode) support for arrays.
- Marked Option<T> types as optional.

### Docs

None

### Description

- Added encoding for arrays following the same format as Vec
- When an Option<T> is None, currently the field is not written at all but the field is still marked as Required. Changed it to optional.

### CODE REVIEW

**Comment by joe-saronic** [comment]
I recommend adding a test case using a contrived struct with an an array, a Some option, a None option, nested type in the option, option in a nested type, and finally an array of options. Pretty sure you will have to put a restriction on an array of raw options.

File: rust/foxglove/src/protobuf.rs [hunk 1]
```diff
@@ -483,6 +483,66 @@ where
...
+        self.iter().map(|value| value.encoded_len()).sum()
+    }
+
+    fn encoded_len_tagged(&self, field_number: u32) -> usize {
+        // Each element is written with its own tag, so sum up the tagged lengths.
+        self.iter()
+            .map(|value| value.encoded_len_tagged(field_number))
+            .sum()
+    }
+}
```
[the remark this answers is not in the document]
Comment (joe-saronic) [reply rust/foxglove/src/protobuf.rs:559]: Oh nice. The AI caught it

**Review by gasmith — COMMENTED** [review]
Looks reasonable. Please address the bot's comments and add test cases.
